File: 1D PIC/simulation.py
from particle import Particle
from node import Node
from pdfsampler import PdfSampler
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import numpy as np
import copy
import constants as const

class Snapshot:
    def __init__(self, time, particles, nodes):
        self.time = time
        self.particles = particles
        self.nodes = nodes

def get_moments(particles: list[Particle], nodes: list[Node], dx):
    half_dx = 0.5*dx
    moments = np.zeros((3, len(nodes)))
    for i, n in enumerate(nodes):
        moments[0, i] = np.sum(p.weight_factor for p in particles if abs(p.x-n.x) < half_dx)*const.m_electron/dx
        moments[1, i] = np.sum(p.weight_factor*p.v for p in particles if abs(p.x-n.x) < half_dx)*const.m_electron/dx
        moments[2, i] = np.sum(p.weight_factor*p.v**2 for p in particles if abs(p.x-n.x) < half_dx)*const.m_electron/dx
    return moments

class Simulation:
    def __init__(self,
                 n_cells, n_particles_per_cell,
                 dt, end_time,
                 f0, x_domain, v_domain,
                 background_charge_density,
                 particle_order = 1,
                 particle_weight_factor=1.0,
                 snapshot_interval = 10,
                 color_rule = lambda x, v: None):
        self.n_cells = n_cells
        self.n_particles_per_cell = n_particles_per_cell
        self.time = 0
        self.end_time = end_time
        self.dt = dt
        self.x_domain = x_domain
        self.L = x_domain[1] - x_domain[0]
        self.v_domain = v_domain
        self.dx = (x_domain[1] - x_domain[0]) / n_cells
        x, v = PdfSampler.sample(f0, [x_domain, v_domain], n_particles_per_cell*n_cells)
        self.particles = [Particle(pos, vel, x_domain, spline_order=particle_order, weight_factor=particle_weight_factor, color=color_rule(pos, vel)) for pos, vel in zip(x, v)]
        self.nodes = [Node(x) for x in np.arange(x_domain[0]+0.5*self.dx, x_domain[1], self.dx)]
        self.background_charge_density = background_charge_density
        self.snapshot_interval = snapshot_interval
        self.snapshots = []
        self.save_snapshot()

    def do_particle_motion(self):
        for p in self.particles:
            p.x_true += p.v * self.dt
            if p.x > self.x_domain[1]:
                p.x -= self.L
            if p.x < self.x_domain[0]:
                p.x += self.L
            for n in self.nodes:
                A = p.get_area_fraction(n.x, self.dx)
                if A > 0:
                    n.particles.append((p, A))
                elif n == self.nodes[0]:
                    A = p.get_area_fraction(n.x+self.L, self.dx)
                    if A > 0:
                        n.particles.append((p, A))
                elif n == self.nodes[-1]:
                    A = p.get_area_fraction(n.x-self.L, self.dx)
                    if A > 0:
                        n.particles.append((p, A))

    def do_charge_assignment(self):
        for n in self.nodes:
            n.charge_density = const.q_electron/self.dx * sum(p.weight_factor*A for p, A in n.particles)
            n.charge_density += self.background_charge_density

    def do_electrostatic_potential(self):
        A = np.zeros((self.n_cells, self.n_cells))
        for i in range(self.n_cells):
            A[i, i-1] = 1
            A[i, i] = -2
            if i+1 < self.n_cells:
                A[i, i+1] = 1
            else:
                A[i, 0] = 1
                
        b = np.fromiter((-self.dx**2*n.charge_density/const.epsilon0 for n in self.nodes), dtype=float)
        phi = np.linalg.pinv(A) @ b
        for i, n in enumerate(self.nodes):
            n.electrostatic_potential = phi[i]

    def do_electric_field_evolution(self):
        for i in range(self.n_cells-1):
            self.nodes[i].electric_field = -0.5*(self.nodes[i+1].electrostatic_potential - self.nodes[i-1].electrostatic_potential)/self.dx
        self.nodes[-1].electric_field = -0.5*(self.nodes[0].electrostatic_potential - self.nodes[-2].electrostatic_potential)/self.dx

    def do_force_assignment(self):
        for p in self.particles:
            p.electric_field = 0
        for n in self.nodes:
            for p, A in n.particles:
                p.electric_field += n.electric_field*A
            n.particles.clear()

    def do_particle_acceleration(self):
        for p in self.particles:
            p.v += const.q_electron/const.m_electron * p.electric_field*self.dt

    def update(self):
        self.do_particle_motion()
        self.do_charge_assignment()
        self.do_electrostatic_potential()
        self.do_electric_field_evolution()
        self.do_force_assignment()
        self.do_particle_acceleration()
        self.time += self.dt

    def run(self, save_snapshots=True):
        while self.time < self.end_time:
            self.update()
            if round(self.time / self.dt) % self.snapshot_interval == 0:
                print(f't = {self.time:.3g}/{self.end_time:.3g} ({self.time/self.end_time:.2%})')
                if save_snapshots:
                    self.save_snapshot()

    def save_snapshot(self):
        self.snapshots.append(Snapshot(self.time, copy.deepcopy(self.particles), copy.deepcopy(self.nodes)))

    def show_snapshots(self, fps=10, save_animation=False, filename='PIC_simulation', repeat=True, show_moments=True, show_cells=False):
        print('Generating animation...')
        if show_moments:
            n_plots = 4
        else:
            n_plots = 1
        fig, ax = plt.subplots(n_plots, 1, figsize=(10, n_plots*5), sharex=True)
        colors = np.unique([p.color for p in self.particles])
        if show_moments:
            moment_array = np.array([get_moments(s.particles, s.nodes, self.dx) for s in self.snapshots])

        def show(s):
            if not show_moments:
                ax.clear()
                for color in colors:
                    x, v = np.array([[p.x, p.v] for p in s.particles if p.color == color]).T
                    ax.scatter(x, v, alpha=0.5, color=color)
                ax.set_title(f'Time = {s.time: .3g}')
                ax.set_ylabel('$v_x$ (m/s)')
                ax.set_ylim(self.v_domain)
                ax.set_xlim(self.x_domain)
                ax.set_xlabel('$x$ (m)')
            else:
                ax[0].clear()
                for color in colors:
                    x, v = np.array([[p.x, p.v] for p in s.particles if p.color == color]).T
                    ax[0].scatter(x, v, alpha=0.5, color=color)
                ax[0].set_title(f'Time = {s.time: .3g}')
                ax[0].set_ylabel('$v_x$ (m/s)')
                ax[0].set_ylim(self.v_domain)

                moments = get_moments(s.particles, s.nodes, self.dx)
                x = np.array([n.x for n in s.nodes])

                for i in range(3):
                    ax[i+1].clear()
                    ax[i+1].plot(x, moments[i])
                    ax[i+1].set_ylabel(f'Moment {i}')
                    ax[i+1].set_ylim([moment_array[:, i, :].min(), moment_array[:, i, :].max()])

                ax[3].set_xlim(self.x_domain)
                ax[3].set_xlabel('$x$ (m)')

            if show_cells:
                for n in self.nodes:
                    for a in ax:
                        a.vlines(n.x-0.5*self.dx, *self.v_domain, linestyles='--', color='k')

        show(self.snapshots[0])
        ani = animation.FuncAnimation(fig=fig, func=show, frames=self.snapshots, interval=1e3/fps, repeat=repeat)
        if save_animation:
            print('Saving...')
            writer = animation.PillowWriter(fps=fps)
            ani.save(f'{filename}.gif', writer=writer)
        print('Displaying...')
        plt.show()
        print('Done!')

    def save_snapshots_to_csv(self, filename):
        print(f'Saving to {filename}...')
        particle_list = []
        node_list = []
        for s in self.snapshots:
            particle_list.append(np.array([[p.x_true, p.v, p.electric_field] for p in s.particles]).flatten('F'))
            node_list.append(np.array([[n.charge_density, n.electric_field] for n in s.nodes]).flatten('F'))
        particle_array = np.vstack(particle_list).T
        node_array = np.vstack(node_list).T
        node_positions = np.fromiter((n.x for n in self.nodes), dtype=float)
        np.savetxt(f'{filename}_particles.csv', particle_array, delimiter=',')
        np.savetxt(f'{filename}_nodes.csv', node_array, delimiter=',')
        np.savetxt(f'{filename}_node_positions.csv', node_positions, delimiter=',')
        print('Done!')

File: Julia/src/simulation.jl
# First, include your custom module files.
# Julia will look for these files in the same directory.
include("constants.jl")
include("pdfsampler.jl")
include("filters.jl")

# -----------------------------------------------------------------------------
# PIC Simulation Module
# -----------------------------------------------------------------------------
module PICSim

# Import necessary packages
using Plots
using Printf
using SparseArrays
using SpecialFunctions
using LinearAlgebra
using DelimitedFiles
using Trapz
using Statistics

# Import the modules we included in the main script.
# The `..` tells Julia to look in the parent scope.
using ..PdfSampler
using ..Constants
import ..Filters

# --- Core Simulation Structs ---

# Struct to hold a single snapshot of the simulation state
struct Snapshot
    time::Float64
    x::Vector{Float64}
    v::Vector{Float64}
    phi::Vector{Float64}
end

# Mutable struct to hold the entire simulation's state and parameters
mutable struct Simulation{SF <: Filters.Filter}
    # Parameters
    n_nodes::Int
    n_particles::Int
    end_time::Float64
    dt::Float64
    x_domain::Tuple{Float64, Float64}
    v_domain::Tuple{Float64, Float64}
    L::Float64
    dx::Float64
    weight_factor::Float64
    shape_function::SF
    snapshot_interval::Int
    
    # State Variables
    time::Float64
    px::Vector{Float64}
    pv::Vector{Float64}
    colors::Vector{String}
    node_positions::Vector{Float64}
    bg_charge_density::Vector{Float64}
    snapshots::Vector{Snapshot}

    # Internal fields (initialized during run)
    shape_function_width::Int
    n_idx_tiling::Vector{Int}
    p_idx::Vector{Int}
    phi_matrix::Matrix{Float64}
    electric_field_matrix::Matrix{Float64}

    # Fields updated every timestep
    interpolation::SparseMatrixCSC{Float64, Int}
    nrho::Vector{Float64}
    ne_field::Vector{Float64}
    pe_field::Vector{Float64}

    # Constructor function
    function Simulation(;
            n_nodes::Int, n_particles_per_cell::Int,
            dt::Float64, end_time::Float64,
            f0::Function, x_domain, v_domain,
            background_charge_density,
            particle_weight_factor::Float64,
            particle_shape_function::SF,
            snapshot_interval::Int=10,
            color_rule=nothing) where {SF <: Filters.Filter}
        
        n_particles = n_particles_per_cell * n_nodes
        L = x_domain[2] - x_domain[1]
        dx = L / n_nodes
        xv = PdfSampler.sample(f0; ranges=[x_domain, v_domain], n_samples=n_particles)
        px = xv[1, :]
        pv = xv[2, :]

        colors = if color_rule !== nothing
            color_rule(px, pv)
        else
            ["cyan" for _ in 1:n_particles]
        end

        node_positions = collect(range(x_domain[1] + 0.5*dx, stop=x_domain[2] - 0.5*dx, length=n_nodes))
        bg_charge_density_vec = background_charge_density(node_positions)

        # Initialize fields with placeholder values
        new{SF}(n_nodes, n_particles, end_time, dt, x_domain, v_domain, L, dx,
            particle_weight_factor, particle_shape_function, snapshot_interval,
            0.0, px, pv, colors, node_positions, bg_charge_density_vec, [],
            0, Int[], Int[], zeros(n_nodes, n_nodes), zeros(n_nodes, n_nodes), # internal fields
            spzeros(n_nodes, n_particles), zeros(n_nodes), zeros(n_nodes), zeros(n_particles) # loop fields
        )
    end
end

# --- Helper Functions ---

# Periodically shift particle positions into the main domain [x_min, x_max)
function shift_x_to_domain(sim::Simulation, x)
    return @. mod((x - sim.x_domain[1]), sim.L) + sim.x_domain[1]
end

# Calculate the shortest periodic distance between a particle and a node
function get_distance(sim::Simulation, px, nx)
    px_domain = shift_x_to_domain(sim, px)
    return @. 0.5 * sim.L - abs(abs(px_domain - nx) - 0.5 * sim.L)
end

# --- Core Physics and Update Functions ---

# Construct the sparse matrix that maps particle quantities to grid nodes
function get_interpolation_matrix(sim::Simulation, x::Vector{Float64})
    # Calculate the nearest node index (1-based) for each particle
    base_n_idx = floor.(Int, ((x .- sim.x_domain[1]) .% sim.L) ./ sim.dx) .+ 1

    # Tile and repeat indices to account for shape function width
    n_idx = repeat(base_n_idx, inner=sim.shape_function_width) .+ sim.n_idx_tiling
    n_idx = mod1.(n_idx, sim.n_nodes) # mod1 for 1-based periodic indexing

    interp_vals = sim.shape_function.(get_distance(sim, x[sim.p_idx], sim.node_positions[n_idx]) ./ sim.dx)
    
    return sparse(n_idx, sim.p_idx, interp_vals, sim.n_nodes, sim.n_particles)
end

# Calculate fluid moments (density, momentum, energy) on the grid
function get_moments(sim::Simulation, px::Vector{Float64}, pv::Vector{Float64})
    px_shifted = shift_x_to_domain(sim, px)
    moments = zeros(3, sim.n_nodes)
    interpolation = get_interpolation_matrix(sim, px_shifted)

    # Note: sum over particles is `dims=2`
    moments[1, :] = sum(interpolation, dims=2)[:] .* (Constants.m_electron * sim.weight_factor / sim.dx)
    moments[2, :] = (interpolation * pv) .* (Constants.m_electron * sim.weight_factor / sim.dx)
    moments[3, :] = (interpolation * (pv .^ 2)) .* (0.5 * Constants.m_electron * sim.weight_factor / sim.dx)
    return moments
end

# The following functions (ending in '!') modify the simulation state `sim`
function push_particles!(sim::Simulation)
    sim.px .+= sim.pv .* sim.dt
end

function interpolate_particles_to_field!(sim::Simulation)
    sim.interpolation = get_interpolation_matrix(sim, sim.px)
    particle_charge_density = sum(sim.interpolation, dims=2)[:]
    sim.nrho = -Constants.q_electron * sim.weight_factor .* particle_charge_density ./ sim.dx^3 .+ sim.bg_charge_density
end

function update_electric_field!(sim::Simulation)
    mul!(sim.ne_field, sim.electric_field_matrix, sim.nrho)
end

function interpolate_field_to_particles!(sim::Simulation)
    mul!(sim.pe_field, transpose(sim.interpolation), sim.ne_field)
end

function accelerate_particles!(sim::Simulation)
    sim.pv .+= -Constants.q_over_m * sim.dt .* sim.pe_field
end

function update!(sim::Simulation)
    push_particles!(sim)
    interpolate_particles_to_field!(sim)
    update_electric_field!(sim)
    interpolate_field_to_particles!(sim)
    accelerate_particles!(sim)
    sim.time += sim.dt
end

function save_snapshot!(sim::Simulation)
    phi = sim.phi_matrix * sim.nrho
    snapshot = Snapshot(sim.time, copy(sim.px), copy(sim.pv), phi)
    push!(sim.snapshots, snapshot)
end

# --- Main Simulation Runner ---

function run!(sim::Simulation; save_snapshots=true)
    # --- One-time setup ---
    y1 = sim.shape_function.(range(-100, 101))
    width = count(x -> abs(x) > 1e-9, y1)
    
    relative_n_idx = (-(width÷2)):(width÷2)
    sim.shape_function_width = length(relative_n_idx)
    sim.n_idx_tiling = repeat(collect(relative_n_idx), sim.n_particles)
    sim.p_idx = repeat(1:sim.n_particles, inner=sim.shape_function_width)

    # Setup finite-difference operators with periodic boundary conditions
    # A is the discrete Laplacian (d^2/dx^2)
    A = spdiagm(
        -1 => ones(sim.n_nodes - 1),
         0 => -2 * ones(sim.n_nodes),
         1 => ones(sim.n_nodes - 1)
    )
    A[1, sim.n_nodes] = 1
    A[sim.n_nodes, 1] = 1
    
    # B is the discrete gradient (-d/dx)
    B = spdiagm(
        -1 => -1 * ones(sim.n_nodes - 1),
         1 =>  1 * ones(sim.n_nodes - 1)
    )
    B[1, sim.n_nodes] = -1
    B[sim.n_nodes, 1] = 1
    
    # Calculate matrices to get potential (phi) and E-field from charge density (rho)
    sim.phi_matrix = -sim.dx^2 * pinv(Matrix(A)) / Constants.epsilon0
    sim.electric_field_matrix = -(0.5 / sim.dx) * B * sim.phi_matrix
    
    # --- Main Loop ---
    println("Starting simulation...")
    start_time = time()
    while sim.time < sim.end_time
        update!(sim)
        current_step = round(Int, sim.time / sim.dt)
        if current_step % sim.snapshot_interval == 0
            percentage = sim.time / sim.end_time * 100
            @printf("t = %.3g/%.3g (%.2f%%)\n", sim.time, sim.end_time, percentage)
            if save_snapshots
                save_snapshot!(sim)
            end
        end
    end
    elapsed = time() - start_time
    @printf("Simulation finished. Elapsed time: %.4f seconds\n", elapsed)
end

# --- Visualization and Export Functions ---

function show_snapshots(sim::Simulation; fps=10, save_animation=false, filename="PIC_simulation", show_moments=true)
    println("Generating animation...")
    
    moment_array = nothing
    if show_moments
        # Pre-calculate all moments for stable y-axis limits
        all_moments = [get_moments(sim, s.x, s.v) for s in sim.snapshots]
        moment_array = permutedims(cat(all_moments..., dims=3), (1, 2, 3))
    end

    unique_colors = unique(sim.colors)

    anim = @animate for (t_idx, s) in enumerate(sim.snapshots)
        if !show_moments
            p = plot(title=@sprintf("Time = %.3g", s.time),
                     xlabel="x (m)", ylabel="\$v_x\$ (m/s)",
                     xlims=sim.x_domain, ylims=sim.v_domain, legend=false)
            for color in unique_colors
                mask = sim.colors .== color
                if any(mask)
                    scatter!(p, shift_x_to_domain(sim, s.x[mask]), s.v[mask], alpha=0.5, color=color, markersize=3)
                end
            end
            plot(p) # Required for @animate
        else
            p1 = plot(title=@sprintf("Time = %.3g", s.time), ylabel="\$v_x\$ (m/s)", ylims=sim.v_domain, legend=false)
            for color in unique_colors
                mask = sim.colors .== color
                if any(mask)
                     scatter!(p1, shift_x_to_domain(sim, s.x[mask]), s.v[mask], alpha=0.5, color=color, markersize=2)
                end
            end

            moments = moment_array[:, :, t_idx]
            p2 = plot(sim.node_positions, moments[1, :], ylabel="Density", ylims=extrema(moment_array[1,:,:]), legend=false)
            p3 = plot(sim.node_positions, moments[2, :], ylabel="Momentum", ylims=extrema(moment_array[2,:,:]), legend=false)
            p4 = plot(sim.node_positions, moments[3, :], ylabel="Energy", ylims=extrema(moment_array[3,:,:]), xlabel="x (m)", legend=false)
            
            plot(p1, p2, p3, p4, layout=(4, 1), size=(800, 1000), sharex=true, xlims=sim.x_domain)
        end
    end

    if save_animation
        println("Saving animation to $filename...")
        gif(anim, "$(filename).gif", fps=fps)
    end
    println("Done!")
    return anim # Return animation object for display in environments like Pluto/Jupyter
end

function save_snapshots_to_csv(sim::Simulation, filename_base::String)
    println("Saving snapshots to $(filename_base)*.csv...")
    particle_data = hcat([vcat(s.x, s.v) for s in sim.snapshots]...)
    writedlm("$(filename_base)_particles.csv", particle_data, ',')
    writedlm("$(filename_base)_node_positions.csv", sim.node_positions, ',')
    println("Done!")
end

end # --- End of PICSim Module ---